pinot-core/src/main/java/com/linkedin/pinot/server/realtime/ControllerLeaderLocator.java

```diff
@@ -69,18 +71,24 @@ public static ControllerLeaderLocator getInstance() {
   /**
    * Locate the controller leader so that we can send LLC segment completion requests to it.
    *
-   * @param forceRefresh : If set to true, then makes a zk call to re-fetch the leadership information.
    * @return The host:port string of the current controller leader.
    */
-  public String getControllerLeader(boolean forceRefresh) {
-    if (controllerLeaderHostPort == null || forceRefresh) {
-      BaseDataAccessor<ZNRecord> dataAccessor = _helixManager.getHelixDataAccessor().getBaseDataAccessor();
-      Stat stat = new Stat();
-      ZNRecord znRecord = dataAccessor.get("/" + _clusterName + "/CONTROLLER/LEADER", stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
-      String leader = znRecord.getId();
-      int index = leader.lastIndexOf('_');
-      controllerLeaderHostPort = leader.substring(0, index) + ":" + leader.substring(index+1);
+  public synchronized String getControllerLeader() {
+    if (!_refresh) {
+      return _controllerLeaderHostPort;
     }
-    return controllerLeaderHostPort;
+
+    BaseDataAccessor<ZNRecord> dataAccessor = _helixManager.getHelixDataAccessor().getBaseDataAccessor();
+    Stat stat = new Stat();
+    ZNRecord znRecord = dataAccessor.get("/" + _clusterName + "/CONTROLLER/LEADER", stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+    String leader = znRecord.getId();
+    int index = leader.lastIndexOf('_');
+    _controllerLeaderHostPort = leader.substring(0, index) + ":" + leader.substring(index + 1);
+    _refresh = false;
+    return _controllerLeaderHostPort;
+  }
+
+  public synchronized void refreshControllerLeader() {
```

Comment:
No, for 2 reasons:
1. I want to minimize zk watches (as yet, it is 0 for LL segments)
2. We still have to handle NOT_LEADER and back off and retry because of the race condition.